contracts/TokenManager.sol

```diff
@@ -69,49 +77,63 @@ contract TokenManager is Ownable, Modifiable, SelfDestructible {
         emit DeregisteredTokenController(_id);
     }
 
-    function registerToken(address token, uint32 interface_id) external onlyOwner notNullAddress(token) {
-        require(interface_id >= 1 && interface_id <= 1048576);
-        require(registeredTokens[token] == 0);
+    function registerToken(address token, string standard) external onlyOwner notNullAddress(token) {
+        require(bytes(standard).length > 0);
+        bytes32 _id = keccak256(abi.encodePacked(standard));
+
+        require(registeredTokens[token].interface_id == bytes32(0));
 
-        registeredTokens[token] = interface_id;
+        registeredTokens[token].interface_id = _id;
 
         //raise event
-        emit RegisteredToken(token, interface_id);
+        emit RegisteredToken(token, _id);
     }
 
     function deregisterToken(address token) external onlyOwner {
-        require(registeredTokens[token] != 0);
+        require(registeredTokens[token].interface_id != 0);
 
-        registeredTokens[token] = 0;
+        registeredTokens[token].interface_id = bytes32(0);
+        registeredTokens[token].blacklisted = false;
 
         //raise event
         emit DeregisteredToken(token);
     }
 
     function blacklistToken(address token) external onlyOwner {
-        require(registeredTokens[token] != 0);
+        require(registeredTokens[token].interface_id != bytes32(0));
 
-        registeredTokens[token] = registeredTokens[token] | BLACKLISTED;
+        registeredTokens[token].blacklisted = true;
 
         //raise event
         emit BlacklistedToken(token);
     }
 
     function whitelistToken(address token) external onlyOwner {
-        require(registeredTokens[token] != 0);
+        require(registeredTokens[token].interface_id != bytes32(0));
 
-        registeredTokens[token] = registeredTokens[token] & (~BLACKLISTED);
+        registeredTokens[token].blacklisted = false;
 
         //raise event
         emit WhitelistedToken(token);
     }
 
-    function getTokenController(address token) public view returns (TokenController) {
-        uint32 interface_id = registeredTokens[token];
+    function getTokenController(address token, string standard) public view returns (TokenController) {
+        if (registeredTokens[token].interface_id != bytes32(0)) {
+            require(!registeredTokens[token].blacklisted);
+
+            address interface_address = registeredTokenControllers[registeredTokens[token].interface_id];
+            require(interface_address != address(0));
+
+            return TokenController(interface_address);
+        }
+
+        if (bytes(standard).length > 0) {
```

Comment:
Here I don't think that any existing mapping should override `standard`. In the event that the mapping of token to interface ID is wrong the given `standard` should take precedence.

Reply:
Fine! I'll switch them